compiler/damlc/lib/DA/Cli/Damlc/DependencyDb.hs

```diff
@@ -175,24 +184,58 @@ dalfFileName bs fp = do
     let pkgId = T.unpack $ LF.unPackageId $ LF.dalfPackageId decodedDalfPkg
     pure $ pkgId </> takeFileName fp
 
-installDataDepDalf :: FilePath -> FilePath -> IO ()
-installDataDepDalf depsDir fp = do
-    bs <- BS.readFile fp
+installDataDepDalf :: Bool -> FilePath -> FilePath -> BS.ByteString -> IO ()
+installDataDepDalf isMain depsDir fp bs = do
     fileName <- dalfFileName bs fp
     let targetFp = depsDir </> fileName
     let targetDir = takeDirectory targetFp
-    markDirWith mainMarker targetDir
+    unlessM (doesDirectoryExist targetDir) $ write targetFp $ BSL.fromStrict bs
+    -- if the directory exists, the dalf got already installed as a dependency
+    when isMain $ markDirWith mainMarker targetDir
     markDirWith dataDepMarker targetDir
-    copy fp targetFp
+
+data DataDeps = DataDeps
+    { dataDepsDars :: [FilePath]
+    , dataDepsDalfs :: [FilePath]
+    , dataDepsPkgIds :: [LF.PackageId]
+    }
+
+readDataDeps :: [String] -> IO DataDeps
+readDataDeps fpOrIds = do
+    pkgIds <- forM pkgIds0 validatePkgId
+    pure $ DataDeps {dataDepsDars = dars, dataDepsDalfs = dalfs, dataDepsPkgIds = pkgIds}
+  where
+    (dars, rest) = partition ("dar" `isExtensionOf`) fpOrIds
+    (dalfs, pkgIds0) = partition ("dalf" `isExtensionOf`) rest
+
+-- | A rudimentary check that no bad package ID's are present in the data-dependency section of
+-- daml.yaml.
+validatePkgId :: String -> IO LF.PackageId
+validatePkgId pkgId = do
+    unless (all (`elem` ['a' .. 'z'] ++ ['A' .. 'Z'] ++ ['0' .. '9']) pkgId) $
```

Comment:
Don’t we store this in base16 with only lowercase letters? Checking that it’s a 64-letter hex string seems easy enough so I’d rather do it properly than use the approximation here.

Reply:
switched to `isBase64` and `all isLower` for the check.

Reply:
why base64? it should be base16 with 64 characters (256 bits because sha256).